xla/service/gpu/nccl_collective_thunk.cc

```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
   return device_buffers;
 }
 
+Status MaybeRegisterBuffers(int device_ordinal,
+                            const std::vector<DeviceBufferPair>& buffers,
+                            ncclComm_t comm) {
+  // TODO(tmorris): Only register if xla_gpu_enable_nccl_user_buffers is true.
+#if XLA_ENABLE_XCCL
+  // Keep track of which communicators we have registered for already.
+  // Each device has one NCCL buffer which only needs to be registered once per
+  // each comm.
+  struct RegisteredBuffers {
+    absl::Mutex mu;
+    absl::flat_hash_map<int, absl::flat_hash_set<ncclComm_t>> per_device_comms
+        ABSL_GUARDED_BY(mu);
+    // Buffers could be deregistered with ncclCommDeregister.
+    std::vector<void*> handles ABSL_GUARDED_BY(mu);
+  };
+  static auto& all_registered = *new RegisteredBuffers;
+
+  auto maybe_register = [&](const se::DeviceMemoryBase& buffer) {
+    if (!all_registered.per_device_comms[device_ordinal].contains(comm)) {
```

Comment:
Thread Safety Analysis at clang can't determine that this lambda is called only with mutex locked. So compilation with clang failing here with error:
error: reading variable 'per_device_comms' requires holding mutex 'all_registered.mu' [-Werror,-Wthread-safety-analysis]

There is an attribute which can disable this check for method:
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-thread-safety-analysis

But unfortunately it doesn't work for lambdas. Can this lambda be extracted to different method if RegisteredBuffers will became singleton?